feat(new sink): Initial `pulsar` sink implementation

leshow opened the pull request:
closes #690 

Missing tests and proper healthcheck right now, I just wanted to show this version to get some feedback. It's based largely on the kafka sink. It impl's `Sink`, like the kafka sink it holds the in flight futures and `Ack`s back after completion. 

Let me know if this approach is suitable, if there are some configuration options missing we'd like to add, etc. I'd like to hold off on SSL at the moment because I don't think it's well supported in the underlying crate. This change depends on a change to `SinkContext` and `pulsar-rs` also, worth looking at that.

binarylogic commented:
@bruceg do you mind reviewing this?

binarylogic commented:
And thanks for doing this @leshow.

bruceg commented on src/sinks/pulsar.rs:
```diff
@@ -0,0 +1,178 @@
+
+fn healthcheck(config: &PulsarSinkConfig, pulsar: &Pulsar) -> crate::Result<super::Healthcheck> {
+    let consumer = pulsar
+        .consumer()
```
I don't know in detail how Pulsar handles authorization, but trying to access the remote as a consumer in `healthcheck` but as a producer in the actual sink could hit different authorization paths.

I see the `kafka` sink also uses this approach, but it is a potential issue to be aware of.

bruceg commented on src/sinks/pulsar.rs:
```diff
@@ -0,0 +1,178 @@
+                    }
+                    self.acker.ack(num_to_ack);
+                }
+                Err(e) => error!("future cancelled: {}", e),
```
This could use a better message. The `kafka` error at least referenced that it was a "delivery" future. The term "future" probably has little meaning to the average Vector user.

bruceg commented:
It looks like a non-issue to me, but I'd feel more comfortable if @LucioFranco or @lukesteensen took a glance at the `SinkContext` change to make sure there wasn't some obvious problem with exposing the executor this way.

bruceg commented:
BTW @leshow you can mark a pull request as "draft" when you submit it by clicking the arrow beside the "Create pull request".

leshow commented:
Thanks for the feedback. Okay, will do for next time, I don't see an option to do it now. So long as the structure is okay I'll wrap up the healthcheck and tests hopefully in the next few days.

KannarFr commented:
Hi, @leshow there is currently no authentication support on that PR, right?

leshow commented:
@KannarFr Correct, however if all you need is the ability to pass the `Authentication` user/pass combo to pulsar I'm pretty sure I can add that in for you. https://docs.rs/pulsar/0.2.0/pulsar/struct.Pulsar.html#method.new

KannarFr commented:
Pulsar is not using user/pass auth but I'm sure it's not difficult, I think
I will commit that PR. :)

Le lun. 3 févr. 2020 à 18:40, Evan Cameron <notifications@github.com> a
écrit :

> @KannarFr <https://github.com/KannarFr> Correct, however if all you need
> is the ability to pass the Authentication user/pass combo to pulsar I'm
> pretty sure I can add that in for you.
> https://docs.rs/pulsar/0.2.0/pulsar/struct.Pulsar.html#method.new
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/timberio/vector/pull/1665?email_source=notifications&email_token=ABZLQJP2LYRSGH7BMWQC2WDRBBJJFA5CNFSM4KONMSCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKUXJUY#issuecomment-581530835>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABZLQJOAMG3WKDL7SGYK72LRBBJJFANCNFSM4KONMSCA>
> .
>

KannarFr commented:
Done in https://github.com/leshow/vector/pull/1.

leshow commented:
I've added healthcheck & an integration test. Tried it locally with pulsar in the background and seems to work. I need to add another `Consumer` in the test still. This branch is also pointing at the master branch of `pulsar-rs`, which could be problematic.

binarylogic commented:
@leshow sounds good. Just re-request a review when you're ready.

leshow commented:
Okay, I've moved this out of WIP. It's working locally, if you want to reproduce, run `pulsar-standalone` in the background, and the integration test with:

`cargo test pulsar --features pulsar-integration-test`

I've added what I think are the relevant configs and `toml` files, but I may have missed something. I was just looking at other PRs and copying what they did.

leshow commented:
Seems like I still need to figure out the testing CI.

binarylogic commented:
@leshow you can run them locally via Docker. See [this README](https://github.com/timberio/vector/blob/master/tests/README.md). Once you get the tests passing we'll review.

KannarFr commented:
@leshow can you get time to do it? I can do it by my side if needed.

leshow commented:
@KannarFr if you'd like to do do the docker bit and submit it as a PR to my fork, sure. Other things have come up that have taken some time. You need to get pulsar-standalone running in a container so the integration tests can run.

leshow commented on Cargo.toml:
```diff
@@ -141,6 +141,7 @@ colored = "1.9"
 warp = "0.1.20"
 evmap = { version = "7", features = ["bytes"] }
 logfmt = "0.0.2"
+pulsar = { git = "https://github.com/wyyerd/pulsar-rs" }
```
Absolutely, I mentioned elsewhere in this thread I'm not sure if you actually want to merge this until they do a release. In the meantime, I can add commit hash.

leshow commented:
I'm at a bit of a loss as to how to add pulsar to the container in order to get these tests to run. I've also had an issue on at least 2 different machines, where running `make test` will consume all available system memory then crash with an error like:

```
Starting vector_elasticsearch_1 ... done
   Compiling vector v0.8.0 (/home/leshow/dev/rust/vector)
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/leshow/.rustup/toolchains/1.41.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.103giqk6wtmhpcv6.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.13obbnmdfykyhc3l.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.14zo4roytzzhj6mi.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.15lz2hk08g91v568.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.16whzk623alwv7i8.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.16y5fyibfga3p1jo.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.19h3e3n1uyqbaszt.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.19kbg2dommqxb82f.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.19kc5nhqnompj6au.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1a7fu4u3qti5a3jv.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1hq4w0rrd58d3n8s.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1kgz424er569b6k.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1l3ufz9yktzay4at.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1l3x534mfyibiaa8.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1ohibu38sjtvp7dj.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1prpmiq1z5kawlhi.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1qzrfow0172vxc85.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1tbyvnu4qugdayz1.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1vvmc14v4jr6rfa2.rcgu.o" "/home/leshow/dev/rust/vector/target/debug/deps/vector-a2c9ee1397164ad8.1vxpu4w64y2rr5y5.rcgu.o" "/home/leshow/dev/rust/vector/tar 

...
```

Has anyone had a problem like this?

binarylogic commented:
@leshow @bruceg will take a look and see what's going on.

bruceg commented:
I just ran `make test`, and while it did gobble up a large amount of memory during compilation, it didn't consume all available memory. How much RAM is in the systems you are testing with, and how many CPU cores/threads? Cargo will use more RAM with more threads, as it tries to fill all CPU resources.

leshow commented:
I don't think that's it, I've got 16 GB of ram and 8 cores (16 threads). I just ran it again and while I wasn't able to get a `scrot` before everything crashed, it looks like I've got 16 processes of `/usr/bin/ld` taking up about 1 GB of ram a piece. The other machine was 12 GB and 6 cores but virtualized Linux on macos.

LucioFranco commented:
@leshow try with no default features, that should disable a lot of our c deps and reduce total mem usage.

```
cargo test --no-default-features --features docker
```

leshow commented:
I did a fresh clone of master, killed and removed all my docker containers, and I'm still not having any luck. Here's a shot of `htop` just before it maxed out my ram

![vector](https://user-images.githubusercontent.com/1128302/74892743-772eba80-5358-11ea-9904-e83b7f69ccbe.png)

bruceg commented:
On my 12-thread box, the build took over 11GB. I wouldn't be surprised if it was able to max out 16GB of RAM on a 16-thread CPU, particularly if it started all the docker-compose instances first. You could try to reduce the number of jobs with `cargo test --jobs 8` or less.

leshow commented:
Yeah, that seemed to do it. Thank you. If anyone could point me in the right direction for the way you'd like the pulsar integration test to run, that would be appreciated. I'm not very familiar with docker. It looks like there's an image we may be able to use https://hub.docker.com/r/apachepulsar/pulsar

bruceg commented:
For integration testing, the pulsar server needs to be set up in `docker-compose.yml`. This can set up a local server that the internal tests can query. We have added a series of features for the integration tests, so we don't run them normally, and put them behind a `docker` feature group in the `Cargo.toml` file. Set up a separate test group that is feature gated. You can see this in, for example, the `kafka sink`.

Hoverbear commented:
@leshow That container image looks great. For the `docker-compose.yml`. If you add that as a service for pulsar like the redis one shown in https://docs.docker.com/compose/gettingstarted/#step-3-define-services-in-a-compose-file and do a bit of poking at the standalone config (https://pulsar.apache.org/docs/en/standalone-docker/) it should be possible to get it working. :)

KannarFr commented:
@leshow can you fix missing tests?

binarylogic commented:
@leshow just double-checking on the status here. If you can get the tests to pass we should be close to merging. Thanks for your work!

leshow commented:
I don't really know what else is required to get this test to run. I've created an entry in `docker-compose` for `pulsar`. I can run `make test` from the root dir, but some other tests I haven't interacted with will still fail before the `pulsar` ones run. If I test with `cargo test pulsar --features docker` will that run with the appropriate containers started? The test returns `ConnectionRefused` which indicates to me that pulsar isn't running.

binarylogic commented:
@bruceg would you mind helping with this so we can get it to the finish line? Thanks.

bruceg commented:
Before running the tests, you will need to start the containers required to support the tests. From what I am seeing in the `docker-compose.yml` changes here, that would mean running `docker-compose up pulsar` in another terminal before starting the tests.

If you want to run all the tests, simply running `make test` will run everything, but this is a large job for just testing your one modification.

Also note, there was a recent issue with the `splunk_hec` sink that caused those tests to fail. If you merge your changes with the current master branch, those tests should pass. You'll probably need to rerun `make generate` as well to make sure the documentation is properly formatted.

Off hand, I'm not seeing why the test is failing on CI here. The same port is specified in the docker config as your test. I see that pulsar has a separate service port for HTTP, should this be the port you are using instead of the broker service port?

leshow commented:
`docker-compose up pulsar`  starts and exits almost immediately:

```
>  vector [master] docker-compose up pulsar
Starting vector_pulsar_1 ... done
Attaching to vector_pulsar_1
vector_pulsar_1 exited with code 0
```

`--verbose` output shows:

```
urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/containers/01e176de6deb172178040c9bacd1ee9f32c17e5a22614c02d55082efca0c5acc/json HTTP/1.1" 200 None
compose.cli.verbose_proxy.proxy_callable: docker inspect_container -> {'AppArmorProfile': 'docker-default',
 'Args': [],
 'Config': {'AttachStderr': False,
            'AttachStdin': False,
            'AttachStdout': False,
            'Cmd': ['bash'],
            'Domainname': '',
            'Entrypoint': None,
            'Env': ['PATH=/usr/local/openjdk-8/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
                    'LANG=C.UTF-8',
...

this request over and over again, the address is notably `localhost:None` 
```

leshow commented:
Thanks for the `command`, that got it running for me, I've also go it running locally and passing the tests now.

bruceg commented:
It looks like you should be able to move this over to the just-released version 0.3.0 of the pulsar crate instead of tracking git. It was released yesterday.

bruceg commented:
It looks like there are just a couple of things left before I do a full review. The biggest is that we require a [developer certificate of origin](https://github.com/apps/dco) on all submissions. Please add a sign-off on all your commits and let us know when you consider this code stable, and I will finish my review.

leshow commented:
I have done the signoff a few times but since I have multiple keys tied to my account the check doesn't pass. I'll try again on my home computer.

leshow commented:
The last thing is the schema,

```
---> The metadata schema is invalid. This means the the resulting
     hash from the `/.meta/**/*.toml` files violates the defined
     schema. Errors include:
     
     * The value at `/sinks/pulsar` failed validation for `/definitions/sink`, reason: `required`
     * The value at `/sinks/pulsar/options/batch_size/max_size` failed validation for `/definitions/field/additionalProperties`, reason: `schema`
     * The value at `/sinks/pulsar/options/auth` failed validation for `/definitions/field`, reason: `required`
```

I have a definition for the sink. I'm not sure what this is asking for.

binarylogic commented:
@leshow I just pushed a fixed for the docs.

binarylogic commented:
@leshow please re-tag @bruceg when you are ready for a final review.

leshow commented:
I think it's all there, one error I get running `make generate` is

```
---> Link `urls.pulsar_sink_source` invalid!
     
       https://github.com/timberio/vector/tree/master/src/sinks/pulsar.rs
     
     Please make sure this path or URL exists.
     
```

This url isn't going to exist until it gets merged to master though, not sure what I can do about it.

binarylogic commented:
@leshow don't worry about that. We don't verify links as a PR check.

bruceg commented on docker-compose.yml:
```diff
@@ -518,6 +519,12 @@ services:
+      command: bin/pulsar standalone
+      ports:
+        - "6650:6650"
+        - "8080:8080"
```
It appears the tests are only accessing port 6650. If port 8080 is not needed for testing, it should probably be left out here.

leshow commented on src/sinks/pulsar.rs:
```diff
@@ -0,0 +1,290 @@
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("creating pulsar producer failed: {}", source))]
+    PulsarSinkFailed { source: pulsar::Error },
```
Sure, I'll update this today

leshow replied:
I see lots of the other sinks and sources using `.context` for a result type that doesn't explicitly include the `BuildError` type. It looks like everything just uses `Box<dyn Error...>` despite the error type being explicitly created. Is that desirable?

bruceg replied:
The `#[derive(Snafu)]` for the error type generates separate top-level structs for each member of the enum. These structs include the requisite traits to allow them used in the `.context` method, which is equivalent to something like: `.map_err(|source| BuildError::PulsarSinkFailed { source })` See [the snafu guide](https://docs.rs/snafu/0.4.3/snafu/guide/index.html) for more details.

The `?` following a `.context` method also calls `.into()` on the error value, which results in it being boxed. There are places where having the error type may be desirable, but there are complications with trait interfaces being generic. We use the box for simplicity.

leshow replied:
I've used snafu in a few projects, just curious why the erased error type was used. Thanks.

leshow commented on src/sinks/pulsar.rs:
```diff
@@ -0,0 +1,290 @@
+            name: auth.auth_name,
+            data: auth.auth_token.as_bytes().to_vec(),
+        });
+        let pulsar = Pulsar::new(config.address.parse()?, auth, exec).wait()?;
```
Yes it could block. I didn't know of a better way to do this part. the tokio `0.1` executor provides `spawn` for types that return `()`. I'm not sure what else I could really do here given what is available. `PulsarSink::new` itself would have to run asynchronously for this to be made non-blocking. That would mean that `SinkConfig::new` would need to return a future also. Do you have any ideas?

bruceg replied:
Unfortunately, no, just something we'll need to be aware of. Moving forward, we will likely convert these into `async fn` to be able to handle exactly this kind of issue but for now we will just eat the `wait`.

bruceg commented:
@Hoverbear you had requested some changes previously. Are your concerns handled now?

Hoverbear commented:
I think it's good. :)

leshow commented:
Sorry I took so much of your time working through the build issues. If you start moving things to async/await and need some help feel free to ping me, I can pick up a few things.

Hoverbear commented:
Thank you so much @leshow , we are starting to move to that! You can see #1922 for some details about that and @LucioFranco or @MOZGIII may be able to direct your efforts!

bruceg commented:
While looking into the units of the batch size, we discovered that it appears to be disconnected the pulsar crate (it is only implemented for the consumer). Have you tested that the batch size is actually constrained by the option? Did we guess right that it is supposed to be a number of messages?

Obviously if the option is disconnected in the upstream crate, this isn't your problem, but we should not claim to support it then.

leshow commented:
I'll take a look, one sec

bruceg commented:
Thanks a lot for your work on this @leshow! We're happy to add your contribution.

binarylogic commented:
Agree! Thank you for pushing this to the finish line.

KannarFr commented:
Yey! Thanks!

KannarFr commented:
@bruceg do you plan to release with pulsar sink soon?

binarylogic commented:
`0.10.0` should go out next week pending benchmark results on our [test harness](https://github.com/timberio/vector-test-harness)

KannarFr commented:
@binarylogic

Thanks for you answer, any news? Where are benchmark output?

KannarFr commented:
@leshow 

Documentation missing these fields on pulsar sink configuration:

```
  type = "pulsar" # required
  inputs = ["source_pulsar_sink"] # required
```

And, I've got the following error:

```
kannar@pond ~/git/cc/conf/vector % ../../vector/target/debug/vector validate vector.toml                                                                                            
Mar 31 09:46:48.024 ERROR vector: Configuration error: unknown variant `codec`, expected `text` or `json` for key `sinks.pulsar_output`
Mar 31 09:46:48.024 ERROR vector: Failed to parse config file. path="vector.toml"
```

When using the configuration:

```
[sinks.pulsar_output]
  type = "pulsar" # required
  inputs = ["input"] # required
  address = "address" # required
  topic = "topic" # required

  # Auth
  auth.name = "${AUTH_PROVIDER}" # optional, no default
  auth.token = "${AUTH_TOKEN}" # optional, no default

  encoding.codec = "text" # required
```

As I'm running master maybe it's not related to your code.

binarylogic commented:
Thanks. I've fixed the documentation errors, they will be merged today. And we're planning to release `0.9.0` this week. Benchmarks look good.

> Where are benchmark output?

We use to run a variety of tests via https://github.com/timberio/vector-test-harness. All performance data is stored in a public S# bucket as documented in the readme. Hope that helps!

KannarFr commented:
@bill-bateman thanks for your answer.

Another point: address seems to must ip, can it be an hostname?

KannarFr commented:
@binarylogic thanks for your answer.

Another point: address seems to must ip, can it be an hostname?

leshow commented:
@KannarFr have a look at the docs for `pulsar-rs`, the address is passed off to them.

KannarFr commented:
Got it, will contribute to pulsar-rs :).

KannarFr commented:
@binarylogic any news for 0.9.0?

LucioFranco commented:
@KannarFr should be coming very very soon :)

KannarFr commented:
Ahaha, I'm sorry but I need this tag to define the exherbo.org package using the new pulsar sink. Maybe can you release a 0.9.0 with it and a 0.10.0 very very soon?

KannarFr commented:
@LucioFranco ^

LucioFranco commented:
I am not totally sure what the status on 0.10 is, though I think @binarylogic currently working on 0.9 as we speak.
